syntax = "proto2";

package research.simulation;

option java_multiple_files = true;

// Stores the parameters required by the radiation optics library.
// Next id: 6
message RRTMOptics {
  // Path of NetCDF file containing the longwave lookup tables.
  optional string longwave_nc_filepath = 1;
  // Path of NetCDF file containing the shortwave lookup tables.
  optional string shortwave_nc_filepath = 2;
  // Path of NetCDF file containing the cloud longwave lookup table.
  optional string cloud_longwave_nc_filepath = 3;
  // Path of NetCDF file containing the cloud shortwave lookup table.
  optional string cloud_shortwave_nc_filepath = 4;
}

message GrayAtmosphereOptics {
  // Reference surface pressure.
  optional float p0 = 1 [default = 1e5];
  // The ratio of the pressure scale height to the partial-pressure scale height
  // of the infrared absorber.
  optional float alpha = 2 [default = 3.5];
  // Longwave optical depth of the entire gray atmosphere.
  optional float d0_lw = 3;
  // Shortwave optical depth of the entire gray atmosphere.
  optional float d0_sw = 4;
}

message OpticsParameters {
  oneof optics_scheme {
    RRTMOptics rrtm_optics = 1;
    GrayAtmosphereOptics gray_atmosphere_optics = 2;
  }

  // Order of accuracy of face value reconstruction for interpolating pressure
  // or temperature at grid cell boundaries. At least 3-rd order is recommended.
  optional int32 face_interp_scheme_order = 3 [default = 3];
}

message AtmosphericState {
  // Path of NetCDF file containing the volume mixing ratio of relevant gases.
  optional string atmospheric_state_nc_filepath = 1;

  // The following are the Radiative Forcing Model Intercomparison Project
  // (RFMIP) experiment identifiers for the atmosphere state and location from
  // which boundary conditions and volume mixing ratio profiles will be drawn.
  // RFMIP experiment label. Default experiment corresponds to the present-day
  // atmosphere.
  optional int32 rfmip_expt_label = 2 [default = 0];
  // RFMIP column site index, which uniquely identifies a (latitude, longitude,
  // time) triplet. Default site corresponds to coordinates 13.5 N, 298.5 E.
  optional int32 rfmip_site = 3 [default = 9];

  // Setting the following parameters will override those provided by the RFMIP
  // state.
  // Surface emissivity; the same for all bands.
  optional float sfc_emis = 4;
  // Surface albedo; the same for all bands.
  optional float sfc_alb = 5;
  // The solar zenith angle (in radians).
  optional float zenith = 6;
  // The total solar irradiance (in W/m²).
  optional float irrad = 7;
  // The longwave incident flux at the top of the atmosphere (in W/m²).
  optional float toa_flux_lw = 8 [default = 0.0];
}

message RadiativeTransfer {
  enum AlgorithmType {
    // The default option where no algorithm type is specified.
    UNKNOWN = 0;
    // A parallel algorithm that simplifies the radiative transfer equations in
    // a way that results in more independent workloads that can be executed in
    // parallel.
    PARALLEL = 1;
    // A sequential algorithm that executes the back-subsitution computations
    // of the radiative transfer solver one vertical layer at a time.
    SEQUENTIAL = 2;
  }

  // Type of distributed algorithm to be used for the RTE solver.
  optional AlgorithmType algorithm_type = 1 [default = PARALLEL];
  // The optics library that calculates optical depth, single-scattering albedo,
  // and other local optical properties of the layered atmosphere.
  optional OpticsParameters optics = 2;
  // Data about the prevalent atmospheric gases present in the atmosphere as
  // well as boundary conditions for the surface and top of atmosphere.
  optional AtmosphericState atmospheric_state = 3;
  // The time interval between updates of the radiative fluxes (in seconds).
  // Since the RRTMGP model is very computationally expensive, it is highly
  // recommended that the update cycle be no less than 10 minutes. However, too
  // long an update cycle can lead to instabilities (Pauluis and Emanuel 2004,
  // https://doi.org/10.1175/1520-0493(2004)132<0673:NIRFIC>2.0.CO;2).
  optional float update_cycle_seconds = 4 [default = 600.0];
}
